### PR TITLE
LRQA-55106 | master

### PR DESCRIPTION
**Pull request by kyle-miho**
https://issues.liferay.com/browse/LRQA-55106

### CODE REVIEW

**Comment by kyle-miho** [comment]
@shuyangzhou I encountered an issue with https://issues.liferay.com/browse/LPS-105915 (error logged, not NPE though) when shutting down the server on occasion, partially due to the fact that the ServletContext uses a mock, so liferay-display.xml never gets registered to the created test company.

Theres also a race condition i couldnt resolve if re-indexing is on, and the test module is deployed on startup.

Theres also a chance of another race condition occuring if the company is being created while certain Verification processes are being executed at the same time as the test module is activated ```
during startup: 2020-02-20 20:14:01.627 INFO  [main][SystemCheckOSGiCommands:59] System check is enabled. You can run a system check with the command "system:check" in Gogo shell.
2020-02-20 20:14:01.842 INFO  [Portal Dependency Manager Component Executor--1][VerifyProcess:65] Verifying com.liferay.bookmarks.internal.verify.BookmarksServiceVerifyProcess
2020-02-20 20:14:01.842 INFO  [Portal Dependency Manager Component Executor--1][LoggingTimer:83] Starting com.liferay.bookmarks.internal.verify.BookmarksServiceVerifyProcess#updateFolderAssets
2020-02-20 20:14:01.847 INFO  [Portal Dependency Manager Component Executor--1][LoggingTimer:44] Completed com.liferay.bookmarks.internal.verify.BookmarksServiceVerifyProcess#updateFolderAssets in 5 ms
2020-02-20 20:14:01.848 INFO  [Portal Dependency Manager Component Executor--1][LoggingTimer:83] Starting com.liferay.bookmarks.internal.verify.BookmarksServiceVerifyProcess#verifyTree
2020-02-20 20:14:01.866 INFO  [Portal Dependency Manager Component Executor--1][LoggingTimer:44] Completed com.liferay.bookmarks.internal.verify.BookmarksServiceVerifyProcess#verifyTree in 18 ms
2020-02-20 20:14:01.866 INFO  [Portal Dependency Manager Component Executor--1][VerifyProcess:80] Completed verification process com.liferay.bookmarks.internal.verify.BookmarksServiceVerifyProcess in 24 ms
2020-02-20 20:14:08.723 ERROR [Portal Dependency Manager Component Executor--1][PortletLocalServiceImpl:335] Unable to register remote portlet for company 36834 because it does not exist
2020-02-20 20:14:09.098 ERROR [Portal Dependency Manager Component Executor--1][PortletLocalServiceImpl:335] Unable to register remote portlet for company 36834 because it does not exist
2020-02-20 20:14:09.200 ERROR [Portal Dependency Manager Component Executor--1][PortletLocalServiceImpl:335] Unable to register remote portlet for company 36834 because it does not exist
2020-02-20 20:14:09.231 ERROR [Portal Dependency Manager Component Executor--1][PortletLocalServiceImpl:335] Unable to register remote portlet for company 36834 because it does not exist
2020-02-20 20:14:09.298 ERROR [Portal Dependency Manager Component Executor--1][PortletLocalServiceImpl:335] Unable to register remote portlet for company 36834 because it does not exist
2020-02-20 20:14:09.536 ERROR [Portal Dependency Manager Component Executor--1][PortletLocalServiceImpl:335] Unable to register remote portlet for company 36834 because it does not exist
2020-02-20 20:14:09.610 ERROR [Portal Dependency Manager Component Executor--1][PortletLocalServiceImpl:335] Unable to register remote portlet for company 36834 because it does not exist
2020-02-20 20:14:10.261 ERROR [Portal Dependency Manager Component Executor--1][PortletLocalServiceImpl:335] Unable to register remote portlet for company 36834 because it does not exist
2020-02-20 20:14:10.747 INFO  [main][CompanyGenerator:66] asfdsafsfsdfsfsaf
20-Feb-2020 20:14:13.161 INFO [main] org.apache.catalina.startup.Catalina.start Server startup in [118,089] milliseconds
```

However these scenarios would never occur during our poshi tests since its expected to only use the module after portal is fully initialized, so we are unsure if its necessary to fix them since we will never encounter them

cc: @vicnate5

**Comment by shuyangzhou** [comment]
Do you have to use a mock servletcontext? Why not just use the portal servletcontext?

**Comment by kyle-miho** [comment]
@shuyangzhou dont we need a portletId to use for servletContext though? I tried looking for a way to get the actual servletContext but they either req a httpRequest or something like this
```
String rootPortletId = PortletIdCodec.decodePortletName(portletId);

PortletBag portletBag = PortletBagPool.get(rootPortletId);

return portletBag.getServletContext();
```

**Comment by shuyangzhou** [comment]
See https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portal/internal/servlet/MainServlet.java#L1309

**Comment by kyle-miho** [comment]
@shuyangzhou Using the real servletContext didn't help, the main startup problems are that in between when the company generator creates a company and when it initializes itself, some startup processes are being executed such as (portlet initialization, calls webAppPool.get(companyId, xxx) for all companies), but the created company is not yet generated, so depending on the thread execution it may or may not fail. Similar things happen on shutdown. We probably dont want to do something like blocking all the other threads in between company creation and initialization, and it seems like the only way to truly fix it would be to force it to run and finish before any other portal services activate other than core ones (basically create itself the same time the first company is created), or to defer it until there is nothing else initializing in Portal (which could change as time goes on).

In our real usecase for Poshi Functional tests, its not realistic to have the activate / deactivate methods being ran during startup / shutdown so we wouldnt encounter these issues. We will be aware of the limitation and don't think its worth fixing something that could slow down the module more (blocking other threads fix) or waiting on something that can change in the future (waiting on every portal initialization to finish).

@vicnate5

File: modules/test/company-generator/src/main/java/com/liferay/company/generator/CompanyGenerator.java [hunk 1]
```diff
@@ -89,6 +90,21 @@ protected void deactivate() {
 	private static final Log _log = LogFactoryUtil.getLog(
 		CompanyGenerator.class);
 
+	private static final ServiceTracker<ServletContext, ServletContext>
```
Comment (shuyangzhou) [inline modules/test/company-generator/src/main/java/com/liferay/company/generator/CompanyGenerator.java]: @kyle-miho you are not doing this right, see https://github.com/liferay/liferay-portal/blob/master/modules/apps/application-list/application-list-api/src/main/java/com/liferay/application/list/adapter/PortletPanelAppAdapterRegistry.java#L79

Fix it, if still fails for you, include an error stack, and let's talk tomorrow.

**Comment by shuyangzhou** [comment]
Please see above comment and resend.

**Pull request closed without merging**